src/Acornima.Extras/ScopeInfo.cs

```diff
@@ -0,0 +1,53 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Acornima.Ast;
+
+namespace Acornima;
+
+public sealed class ScopeInfo
+{
+    private readonly int _id;
+    private readonly int _currentVarScopeId;
+    private readonly int _currentThisScopeId;
+
+    public ScopeInfo(Node node, in Scope scope, ReadOnlySpan<Scope> scopeStack)
+    {
+        AssociatedNode = node;
+        _id = scope.Id;
+        _currentVarScopeId = scope.CurrentVarScopeIndex == scopeStack.Length ? scope.Id : scopeStack[scope.CurrentVarScopeIndex].Id;
+        _currentThisScopeId = scope.CurrentThisScopeIndex == scopeStack.Length ? scope.Id : scopeStack[scope.CurrentThisScopeIndex].Id;
+        VarNames = scope.VarNames.ToArray();
+        LexicalNames = scope.LexicalNames.ToArray();
+        FunctionNames = scope.FunctionNames.ToArray();
+    }
+
+    public Node AssociatedNode { get; }
+    public Node? AssociatedNodeParent { get; internal set; }
+
+    public string[] VarNames { get; }
+    public string[] LexicalNames { get; }
+    public string[] FunctionNames { get; }
+
+    // These lookups could as well be cached.
+    public ScopeInfo? Parent => FindAncestor(_ => true);
+    public ScopeInfo? VarScope => FindAncestor(scope => scope._id == _currentVarScopeId);
+    public ScopeInfo? ThisScope => FindAncestor(scope => scope._id == _currentThisScopeId);
+
+    private ScopeInfo? FindAncestor(Func<ScopeInfo, bool> predicate)
```

Comment:
Could we make `FindAncestor` public? is there any downside/reason for it to be private?

Reply:
In the new commit I'm experimenting with another approach that sets up parent references during parsing.

The reason for this is that I expect that walking up the scope tree is one of the most frequent operations that consumers will do. So it's probably worth to optimize for this use case, even at the expense of making parsing a bit more costly (of course, only when opting in to scope info tracking).

With this new approach `FindAncestor` is no longer needed internally, so I don't really want to include it in the public API to keep the API surface area small. Consumers who need it can implement it quite easily themselves, e.g. as an extension method.